### PR TITLE
infra: suffix index.js and worker url with md5

### PR DESCRIPTION
**Pull request by stephanwlee**
TensorBoard resources, especially JavaScripts, are not tiny yet they are
not being cached at all by clients. As such, users are downloading large
amount of content every time browser refreshes and it impacts:
- above the fold time load latency
- server utilization

This change mitigates above two problems by giving us an ability to
leverage caching by creating a unique resource url with md5 hash of the
file.

Do note that browsers treat, for example, `index.js` differently than
`index.js?random_query`.

Internal change: cl/351171360
Test CL: cl/351209032

### CODE REVIEW

File: tensorboard/webapp/BUILD
Comment (wchargin) [inline tensorboard/webapp/BUILD:213]: Maybe give this a name that more clearly indicates the difference
compard to `index.html` (which *isn’t* that one contains more or less
binary content)? e.g., `/index.no_resource_digests.html` following our
existing patterns, but I’m open to suggestions.

File: tensorboard/defs/defs.bzl
Comment (wchargin) [inline tensorboard/defs/defs.bzl:153]: Can we use the full word `resource` in this name?

I’m not in love with `tf_resource_digest_suffixer`, because most rules
are named based on what the output is (a binary, a library, a test)
rather than a “«verb»er”. But I don’t immediately have a suggestion, so
no action required.

File: tensorboard/defs/defs.bzl
Comment (wchargin) [inline tensorboard/defs/defs.bzl:174]: Should be `exec_tools`, as this tool is meant to be run on the [platform
on which Bazel actions run, not the one on which Bazel itself runs.][1]
Cf. #4031.

[1]: https://docs.bazel.build/versions/master/platforms.html

File: tensorboard/defs/defs.bzl
Comment (wchargin) [inline tensorboard/defs/defs.bzl:154]: Cool. I’m okay with a new binary and macro for this. The more we have,
the more I consider defining something like a custom Jinja context with
filters for inlining base64 images, computing file digests, transcluding
files, etc. That way we could have fewer ad hoc genrules everywhere.

File: tensorboard/BUILD
Comment (stephanwlee) [reply tensorboard/BUILD:320]: Webfiles checker is incapable of understanding the URL and matches filename against a path with query parameter. Because we are not using Closure Compiler, this is effectively only suppressing the Webfiles checker which I think is good enough.

Comment (wchargin) [reply tensorboard/BUILD:320]: Hmm, okay. It seemed strange to me because I would have guessed that if
the webfiles validator can’t resolve query strings, whatever actually
does the webfile merging will also fail. And if we’re not doing webfile
merging, then I don’t see why the webfiles validator is even being run.
But I’ll defer to you if you think that this is okay.

Might we perhaps at least change to `suppress = ["strictDependencies"]`?

File: tensorboard/defs/defs.bzl
Comment (wchargin) [inline tensorboard/defs/defs.bzl:177]: `for (template_name, label) in resources.items()`?

File: tensorboard/defs/defs.bzl
Comment (wchargin) [inline tensorboard/defs/defs.bzl:170]: Writing `File` here seems a bit confusing, since [`File`] is a Bazel
concept but these values are just strings (this is a macro, not a rule).
Maybe “file” or “label”?

[`File`]: https://docs.bazel.build/versions/master/skylark/lib/File.html

File: tensorboard/defs/resource_digest_suffixer_test.py
Comment (wchargin) [inline tensorboard/defs/resource_digest_suffixer_test.py:40]: We can remove this, I think? The point of `get_temp_dir()` is that it’s
automatically cleaned up.

**Review by wchargin — APPROVED** [review]
Lovely; thanks!

**Comment by stephanwlee** [comment]
Moved the macro to `tensorboard/defs/internal/resource.bzl`.
Rationale: having everything stuffed into defs.bzl or large bzl file makes it harder to do internal transformations so let's break .bzl down?

File: tensorboard/defs/internal/resource.bzl
Comment (wchargin) [inline tensorboard/defs/internal/resource.bzl:21]: s/md5/hash/

**Review by wchargin — COMMENTED** [review]
I don’t entirely understand the internal/external structure—why not move
the contents of `defs/internal/resource.bzl` to `defs/js.bzl`?—but no
major objection.

**Comment by stephanwlee** [comment]
Any folder with word `internal` is regarded as an internal set of rules and it disallows other packages like What-If tool to reach in. `defs/js.bzl` re-exports only what is supposed to be public but in this case, I am not even sure if I want this to be a public rule/macro. Did that address your question?

**Comment by wchargin** [comment]
> Any folder with word `internal` is regarded as an internal set of rules and it disallows other packages like What-If tool to reach in.

Interesting; I didn’t know this. I don’t entirely understand, since we
load `@io_bazel_rules_webtesting//web/internal:platform_http_file.bzl`
in our `third_party/workspace.bzl`. Isn’t that an external package?

I also don’t quite get why it’s valuable to declare this macro in an
internal `.bzl` file given that we re-export it from `js.bzl`. Wouldn’t
the What-If Tool be able to access it through `js.bzl` anyway?

Agreed that it would be nice to avoid random Google packages taking a
dependency on the macro, but my understanding was that [this isn’t
possible][bzlvis]. That said, since the expanded macro depends on
`:resource_digest_suffixer`, does it suffice to limit the visibility of
that binary to `//tensorboard:internal` (which you’ve done)?

[bzlvis]: https://docs.bazel.build/versions/master/visibility.html#visibility-of-bzl-files

**Comment by stephanwlee** [comment]
> Interesting; I didn’t know this. I don’t entirely understand, since we
load @io_bazel_rules_webtesting//web/internal:platform_http_file.bzl
in our third_party/workspace.bzl. Isn’t that an external package?

Technically, it is not a violation at Bazel level but buildozer would scream foul. If you disregard those lint warnings, I guess it has no powers. In other words, it is not Barlog proof.

**Comment by wchargin** [comment]
I see. In any case, like I said, no major objection; please don’t feel
blocked on my behalf.

**Pull request merged**
